helix-lsp/src/lib.rs

```diff
@@ -540,41 +548,45 @@ impl Registry {
     pub fn stop(&mut self, language_config: &LanguageConfiguration) {
         let scope = language_config.scope.clone();
 
-        if let Some((_, client)) = self.inner.remove(&scope) {
-            tokio::spawn(async move {
-                let _ = client.force_shutdown().await;
-            });
+        if let Some(clients) = self.inner.remove(&scope) {
+            for (_, client) in clients {
+                tokio::spawn(async move {
```

Comment:
I was curious if it's better here to have the `for` inside or outside the `spawn`. Any risk of bad things happening from `tokio::spawn`ing too many futures?

Reply:
Compared to spawning threads spawning tokio tasks is very cheap. In fact I have read that tokio is able to scale up to hundreds of thousands of tasks in production servers. From the docs:

> Tasks are light weight. Because tasks are scheduled by the Tokio runtime rather than the operating system, creating new tasks or switching between tasks does not require a context switch and has fairly low overhead. Creating, running, and destroying large numbers of tasks is quite cheap, especially compared to OS threads.

The idea is basically that an async tokio task spends most of its time `await`ing something so having a lot of them has little overhead because they don't actually do anything most of the time. The nice thing about async io is that the task gets automatically woken up (and queued to be run) once the io is finished so until then `tokio` only needs to store that task [^1]. In the worst case (single threaded executer) tokio essentially runs these tasks consequitvely in a loop so that would be about equivalent to moving the for loop insidie the task (because we await anyway).

This case is a classical example of: spawn a task that does very little computation and instead just sends a message and waits for a response so it just felt right to do it this way. In the (very weird/unlikely) case that we have a large number of servers that all timeout while sending responses this approach is somewhat more optimal because the servers are shutdown in parallel instead of consecutively.
 
That said moving the for loop inside the task works too and there is likely little difference either way. I am certain that we will never reach the point where the overhead of spawning more tokio tasks becomes relevant. An editor just doesn't spawn that many tasks compared to a server :D

I did notice that I was inconsistent here and had the for loop inside the spawn in `restart` I moved that outside too so it's atleast consistent and looks less jarring.

[^1]: That means we should make sure that all "heavy" computations (for example diffs, updating the document running TS) computation are not done in tokio tasks but inside a blocking context instead: see [`block_in_place`](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) and [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html)